### PR TITLE
Revisions to lambda expressions topic.

### PR DESCRIPTION
**Pull request by rpetrusha**
# Revisions to lambda expressions topic

## Summary

Extensively revised with new examples and a section on tuples and lambda expressions.

## Suggested Reviewers
@BillWagner @stevehoag @mairaw

### CODE REVIEW

**Comment by mairaw** [comment]
There are some build errors. It would be good to fix those first, otherwise we can't even see it on staging.

**Comment by rpetrusha** [comment]
@mairaw I think I've fixed all the build errors.  A new build should run shortly.

File: docs/csharp/lambda-expressions.md [hunk 1]
```diff
@@ -1,27 +1,148 @@
----
-title: Lambda Expressions
-description: Lambda expressions are shorthand syntax for methods and delegates. Learn when to use them.
-keywords: .NET, .NET Core
-author: dotnet-bot
-manager: wpickett
-ms.date: 06/20/2016
```
Comment (stevehoag) [inline docs/csharp/lambda-expressions.md]: Update the date; also needs a ms.author.

File: docs/csharp/lambda-expressions.md [hunk 2]
```diff
@@ -1,27 +1,148 @@
...
+
+You can easily create lambda expressions and statements that incorporate asynchronous processing by using the [async](language-reference/keywords/async.md) and [await](language-reference/keywords/await.md) keywords. For example, the example calls a `ShowSquares` method that executes asynchronously.
+
+[!CODE [csSnippets.Lambdas](../../samples/snippets/csharp/concepts/lambda-expressions/async1.cs#1)]
+
+For more information about how to create and use async methods, see [Asynchronous programming with async and await](programming-guide/concepts/async/asynchronous-programming-with-async-and-await.md).
+
+## Lambda expressions and tuples ##
+
+Starting with C# 7.0, the C# language provides built-in support for tuples. You can provide a tuple as an argument to a lambda expression, and your lambda expression can also return a tuple. In some cases, the C# compiler uses type inference to determine the types of tuple components.
```
Comment (stevehoag) [inline docs/csharp/lambda-expressions.md]: Link to the Tuples topic?

File: docs/csharp/lambda-expressions.md [hunk 3]
```diff
@@ -1,27 +1,148 @@
...
+
+The delegate can be instantiated with code like the following
+
+[!CODE [csSnippets.Lambdas](../../samples/snippets/csharp/concepts/lambda-expressions/query1.cs#2)]
+
+where `int` is an input parameter, and `bool` is the return value. The return value is always specified in the last type parameter. When the following `Func` delegate is invoked, it returns true or false to indicate whether the input parameter is equal to 5:
+
+[!CODE [csSnippets.Lambdas](../../samples/snippets/csharp/concepts/lambda-expressions/query1.cs#3)]
+
+You can also supply a lambda expression when the argument type is an @System.Linq.Expressions.Expression%601, for example in the standard query operators that are defined in the @System.Linq.Queryable type. When you specify an @System.Linq.Expressions.Expression%601 argument, the lambda is compiled to an expression tree. The following example uses the @System.Linq.Enumerable.Count%60%601(System.Collections.Generic.IEnumerable{t%60%600})) standard query operator.
```
Comment (stevehoag) [inline docs/csharp/lambda-expressions.md]: The last link (@System.Linq.Enumerable.Count) isn't resolving.

File: docs/csharp/lambda-expressions.md [hunk 4]
```diff
@@ -1,27 +1,148 @@
...
+
+- A lambda expression cannot contain a `goto` statement, `break` statement, or `continue` statement that is inside the lambda function if the jump statement’s target is outside the block. It is also an error to have a jump statement outside the lambda function block if the target is inside the block.
+
+## Featured Book Chapter ##
+
+[Delegates, Events, and Lambda Expressions](http://go.microsoft.com/fwlink/?LinkId=195395) in [C# 3.0 Cookbook, Third Edition: More than 250 solutions for C# 3.0 programmers](http://go.microsoft.com/fwlink/?LinkId=195369)
+
+## See Also ##
+
+[LINQ (Language-Integrated Query)](../standard/using-linq.md)
```
Comment (stevehoag) [inline docs/csharp/lambda-expressions.md]: These three links are all showing up on a single line - was that intentional?

File: docs/csharp/lambda-expressions.md [hunk 5]
```diff
@@ -1,27 +1,148 @@
...
+
+- Variables introduced within a lambda expression are not visible in the outer method.
+
+- A lambda expression cannot directly capture a `ref` or `out` parameter from an enclosing method.
+
+- A return statement in a lambda expression does not cause the enclosing method to return.
+
+- A lambda expression cannot contain a `goto` statement, `break` statement, or `continue` statement that is inside the lambda function if the jump statement’s target is outside the block. It is also an error to have a jump statement outside the lambda function block if the target is inside the block.
+
+## Featured Book Chapter ##
```
Comment (stevehoag) [inline docs/csharp/lambda-expressions.md]: Should we still be linking to this? The book was published in 2007.

**Review by stevehoag — COMMENTED** [review]
Looks good - add a few minor comments.

File: docs/csharp/lambda-expressions.md (hunk 1 above)
Comment (rpetrusha) [inline docs/csharp/lambda-expressions.md]: @stevehoag Thanks. I completely forgot to copy the header from the original topic.

File: docs/csharp/lambda-expressions.md [hunk 6]
```diff
@@ -1,27 +1,160 @@
...
+
+[!CODE [csSnippets.Lambdas](../../samples/snippets/csharp/concepts/lambda-expressions/query1.cs#6)]
+
+## Type inference in lambda expressions ##
+
+When writing lambdas, you often do not have to specify a type for the input parameters because the compiler can infer the type based on the lambda body, the parameter’s delegate type, and other factors, as described in the C# Language Specification. For most of the standard query operators, the first input is the type of the elements in the source sequence. So if you are querying an `IEnumerable<Customer>`, then the input variable is inferred to be a `Customer` object, which means you have access to its methods and properties:
+
+[!CODE [csSnippets.Lambdas](../../samples/snippets/csharp/concepts/lambda-expressions/infer1.cs#1)]
+
+The general rules for type inteference for lambdas are:
```
Comment (mairaw) [inline docs/csharp/lambda-expressions.md]: typo: inteference

File: docs/csharp/lambda-expressions.md [hunk 7]
```diff
@@ -1,27 +1,160 @@
...
+
+- The return value of the lambda (if any) must be implicitly convertible to the delegate's return type.
+
+Note that lambda expressions in themselves do not have a type because the common type system has no intrinsic concept of "lambda expression." However, it is sometimes convenient to speak informally of the "type" of a lambda expression. In these cases the type refers to the delegate type or @System.Linq.Expressions.Expression type to which the lambda expression is converted.
+
+## Variable Scope in Lambda Expressions ##
+
+Lambdas can refer to *outer variables* (see [Anonymous methods](programming-guide/statements-expressions-operators/anonymous-methods.md)) that are in scope in the method that defines the lambda function, or in scope in the type that contains the lambda expression. Variables that are captured in this manner are stored for use in the lambda expression even if the variables would otherwise go out of scope and be garbage collected. An outer variable must be definitely assigned before it can be consumed in a lambda expression. The following example demonstrates these rules.
```
Comment (mairaw) [inline docs/csharp/lambda-expressions.md]: must be definitely -> must be (I think the must is enough)

File: docs/csharp/lambda-expressions.md [hunk 8]
```diff
@@ -1,27 +1,160 @@
...
+
+[!CODE [csSnippets.Lambdas](../../samples/snippets/csharp/concepts/lambda-expressions/query1.cs#5)]
+
+The following example specifies multiple input parameters by enclosing them in parentheses. The method returns all the elements in the numbers array until it encounters a number whose value is less than its ordinal position in the array.
+
+[!CODE [csSnippets.Lambdas](../../samples/snippets/csharp/concepts/lambda-expressions/query1.cs#6)]
+
+## Type inference in lambda expressions ##
+
+When writing lambdas, you often do not have to specify a type for the input parameters because the compiler can infer the type based on the lambda body, the parameter’s delegate type, and other factors, as described in the C# Language Specification. For most of the standard query operators, the first input is the type of the elements in the source sequence. So if you are querying an `IEnumerable<Customer>`, then the input variable is inferred to be a `Customer` object, which means you have access to its methods and properties:
```
Comment (mairaw) [inline docs/csharp/lambda-expressions.md]: nit: missing comma after So

File: docs/csharp/lambda-expressions.md [hunk 9]
```diff
@@ -1,27 +1,160 @@
...
-manager: wpickett
-ms.date: 06/20/2016
-ms.topic: article
-ms.prod: visual-studio-dev-14
-ms.technology: devlang-csharp
-ms.devlang: csharp
-ms.assetid: b6a0539a-8ce5-4da7-adcf-44be345a2714
+-title: Lambda Expressions
+-description: Lean to use lambda expressions, which are executable code blocks that can be passed as arguments. 
+-keywords: .NET, .NET Core, lambda epxressions, lambdas, delegates
```
Comment (BillWagner) [inline docs/csharp/lambda-expressions.md]: typo: epxressions

File: docs/csharp/lambda-expressions.md [hunk 10]
```diff
@@ -1,27 +1,160 @@
...
+
+A *lambda expression* is a block of code (an expression or a statement block) that is treated as an object. It can be passed as an argument to methods, and it can also be returned by method calls. Lambda expressions are used extensively for:
+
+- Passing the code that is to be executed to asynchronous methods, such as @System.Threading.Tasks.Task.Run(System.Action).
+
+- Writing [LINQ query expressions](linq.md).
+
+- Creating [expression trees](expression-trees-building.md).
+
+Lambda expressions are delegates. The specific delegate type of a lambda expressions depends on its parameters and return value. Lambda expressions that don't return a value correspond to a specific `Action` delegate, depending on its number of parameters. Lambda expressions that return a value correspond to a specific `Func` delegate, depending on its number of parameters. For example, a lambda expression that has two parameters but returns no value corresponds to an @System.Action%602 delegate. A lambda expression that has one parameter and returns a value corresponds to @System.Func%602 delegate.
```
Comment (BillWagner) [inline docs/csharp/lambda-expressions.md]: Lambda expressions are delegates, except when they are expression trees. 

Maybe "Lambda expressions are code that can be represented either as a delegate, or an expression tree that would compile to a delegate".  (I'm not sure I like that wording, improve at will.)

File: docs/csharp/lambda-expressions.md [hunk 11]
```diff
@@ -1,27 +1,160 @@
...
+
+Two or more input parameters are separated by commas enclosed in parentheses:
+
+[!CODE [csSnippets.Lambdas](../../samples/snippets/csharp/concepts/lambda-expressions/expression3.cs#2)]
+
+Ordinarily, the compiler uses type inference in determining parameter types. However, sometimes it is difficult or impossible for the compiler to infer the input types. When this occurs, you can specify the types explicitly, as in the following example:
+
+[!CODE [csSnippets.Lambdas](../../samples/snippets/csharp/concepts/lambda-expressions/expression3.cs#3)]
+
+Note in the previous example that the body of an expression lambda can consist of a method call. However, if you are creating expression trees that are evaluated outside of the .NET Framework, such as in SQL Server, you should not use method calls in lambda expressions. The methods will have no meaning outside the context of the .NET runtime.
```
Comment (BillWagner) [inline docs/csharp/lambda-expressions.md]: We should add some weasel words here. Entity Framework does translate some known methods (especially on the string class. The guidance is correct in the general case.

File: docs/csharp/lambda-expressions.md [hunk 12]
```diff
@@ -1,27 +1,160 @@
...
+
+```cs
+(input parameters) => { statement; }
+```
+
+The body of a statement lambda can consist of any number of statements; however, in practice there are typically no more than two or three.
+
+[!CODE [csSnippets.Lambdas](../../samples/snippets/csharp/concepts/lambda-expressions/statement1.cs#1)]
+
+Statement lambdas, like anonymous methods, cannot be used to create expression trees.
```
Comment (BillWagner) [inline docs/csharp/lambda-expressions.md]: I'd recommend deleting "like anonymous methods".  The C# spec defines lambda expressions as a syntax for anonymous methods, so in that sense all lambda expressions are anonymous methods.

File: docs/csharp/lambda-expressions.md [hunk 13]
```diff
@@ -1,27 +1,160 @@
...
+
+Ordinarily, the fields of a tuple are named `Item1`, `Item2`, etc. You can, however, define a tuple with named components, as the following example does.
+
+[!CODE [csSnippets.Lambdas](../../samples/snippets/csharp/concepts/lambda-expressions/tuples2.cs#1)]
+
+For more information on support for tuples in C#, see [C# Tuple types](tuples.md).
+
+## Lambdas with the standard query operators ##
+
+Many standard query operators have an input parameter whose type is one of the @System.Func%601 family of generic delegates. These delegates use type parameters to define the number and type of input parameters, and the return type of the delegate. `Func` delegates are very useful for encapsulating user-defined expressions that are applied to each element in a set of source data. For example, consider the @System.Func%601 delegate, whose syntax is:
```
Comment (BillWagner) [inline docs/csharp/lambda-expressions.md]: I would start this by saying "LINQ to Objects, among other implementations" have an input parameter..

That way this is distinguished from the IQueryable discussion later.

File: docs/csharp/lambda-expressions.md (hunk 8 above)
Comment (BillWagner) [inline docs/csharp/lambda-expressions.md]: I'd just remove "So".

Comment (BillWagner) [inline docs/csharp/lambda-expressions.md]: I think you should delete "delegate" in the phrase "the parameter's delegate type". Or am I misunderstanding this sentence?

File: docs/csharp/lambda-expressions.md (hunk 7 above)
Comment (BillWagner) [inline docs/csharp/lambda-expressions.md]: No, this should stay "definitely assigned". That's the phrase the spec always uses for the analysis of assignment to a variable before it is used.

**Review by BillWagner — CHANGES_REQUESTED** [review]
Ron, I like the way you've explained all the concepts here.

I had a few locations where the language is technically incorrect based on the terms used in the spec.  once those are handled, this looks great.

File: docs/csharp/lambda-expressions.md (hunk 6 above)
Comment (rpetrusha) [inline docs/csharp/lambda-expressions.md]: @mairaw And I didn't even spell the wrong word (interference) correctly. Fixed.

**Comment by rpetrusha** [comment]
@BillWagner I've addressed all of your comments and will update the topic.

File: docs/csharp/lambda-expressions.md (hunk 7 above)
Comment (mairaw) [inline docs/csharp/lambda-expressions.md]: Good to know @BillWagner 😄

**Review by BillWagner — APPROVED** [review]

**Comment by BillWagner** [comment]
@rpetrusha LGTM.  :shipit:

**Pull request merged**